presto-main/src/main/java/com/facebook/presto/cost/HistoricalPlanStatisticsUtil.java

```diff
@@ -0,0 +1,167 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.cost;
+
+import com.facebook.presto.spi.statistics.HistoricalPlanStatistics;
+import com.facebook.presto.spi.statistics.PlanStatistics;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import static java.lang.Double.isNaN;
+
+public class HistoricalPlanStatisticsUtil
+{
+    // Store historical data for last ``MAX_LAST_RUNS`` runs at maximum.
+    private static final int MAX_LAST_RUNS = 10;
+    // Maximum number of input tables till we store data of multiple last runs. Once number of input tables becomes
+    // larger than this, we store data of fewer historical runs. See getMaxLastRuns() function.
+    private static final int MAX_INPUT_TABLES = 100;
+    // Use a historical statistic when size of input tables in within 1 +/- ``THRESHOLD`` size range.
+    private static final double THRESHOLD = 0.1;
+
+    private HistoricalPlanStatisticsUtil() {}
+
+    /**
+     * Returns predicted plan statistics depending on historical runs
+     */
+    public static PlanStatistics getPredictedPlanStatistics(
+            HistoricalPlanStatistics historicalPlanStatistics,
+            List<PlanStatistics> inputTableStatistics)
+    {
+        Optional<PlanStatistics> overriddenStatistics = historicalPlanStatistics.getOverriddenStatistics();
+        Optional<List<PlanStatistics>> lastRunsStatistics = historicalPlanStatistics.getLastRunsStatistics();
+        Optional<List<List<PlanStatistics>>> lastRunsInputStatistics = historicalPlanStatistics.getLastRunsInputStatistics();
+
+        if (overriddenStatistics.isPresent()) {
+            return overriddenStatistics.get();
+        }
+
+        if (!lastRunsStatistics.isPresent() || !lastRunsInputStatistics.isPresent()) {
+            return PlanStatistics.empty();
+        }
+
+        if (lastRunsInputStatistics.get().size() != lastRunsStatistics.get().size() ||
+                lastRunsStatistics.get().isEmpty() ||
+                lastRunsInputStatistics.get().get(0).size() != inputTableStatistics.size()) {
+            return PlanStatistics.empty();
+        }
+
+        Optional<Integer> similarStatsIndex = getSimilarStatsIndex(historicalPlanStatistics, inputTableStatistics);
+
+        if (similarStatsIndex.isPresent()) {
+            return lastRunsStatistics.get().get(similarStatsIndex.get());
+        }
+
+        // TODO: Use linear regression to predict stats if we have only 1 table.
+        return PlanStatistics.empty();
+    }
+
+    /**
+     * Returns updated HistoricalPlanStatistics after an update with new PlanStatistics
+     */
+    public static HistoricalPlanStatistics trackPlanStatistics(
```

Comment:
can we name this and trackStatistics something else?  For some reason I find it very confusing, and every time I need to look into the code to see what trackStatistics does.